### PR TITLE
Add container snapshot expiry

### LINKED ISSUE #5387: Scheduled snapshot expiry
I saw in [this post](url
https://github.com/lxc/lxd/issues/2610) that there may be interest in snapshot expiry. The thread said 

> If there is interest for this part specifically, please file a separate issue about it and we'll schedule it to be implemented after this one.

So here I am, filing a separate issue about it because I didn't see another one. I don't need hourly snapshots of my container going back into the last week, only the last day. And I only need weekly snapshots into the last 2 weeks, etc.

### PR DESCRIPTION
**Pull request by monstermunchkin**
Resolves #5387

### CODE REVIEW

File: shared/api/container_snapshot.go [hunk 1]
```diff
@@ -20,8 +20,18 @@ type ContainerSnapshotPost struct {
 	Live bool `json:"live,omitempty" yaml:"live,omitempty"`
 }
 
+// ContainerSnapshotPut represents the modifiable fields of a LXD container snapshot
+// API extension: snapshot_expiry
+type ContainerSnapshotPut struct {
+	// Using a pointer here allows us to differentiate between missing key, empty value,
+	// and provided value. This is important when editing the snapshot config.
+	ExpiresAt *time.Time `json:"expires_at" yaml:"expires_at"`
```
Comment (stgraber) [inline shared/api/container_snapshot.go]: We should find a way to do this without using a pointer as that'd be inconsistent with what we do for the rest of the timestamps in the API.

File: shared/api/container.go [hunk 2]
```diff
@@ -55,6 +55,11 @@ type ContainerPut struct {
 
 	// API extension: entity_description
 	Description string `json:"description" yaml:"description"`
+
+	// API extension: snapshot_expiry
+	// This field is used for `lxc config show` only, and is only shown for snapshots
+	// since regular containers cannot expire.
+	ExpiresAt *time.Time `json:"expires_at,omitempty" yaml:"expires_at,omitempty"`
```
Comment (stgraber) [inline shared/api/container.go]: I think I'd prefer we don't do this kind of trickery and instead change `lxc config show` to not abuse ContainerPut for snapshots.

File: shared/container.go [hunk 3]
```diff
@@ -285,6 +285,7 @@ var KnownContainerConfigKeys = map[string]func(value string) error{
 	},
 	"snapshots.schedule.stopped": IsBool,
 	"snapshots.pattern":          IsAny,
+	"snapshots.expire":           IsBool,
```
Comment (stgraber) [inline shared/container.go]: Shouldn't this be `snapshots.expiry` and take some kind of delay (probably supporting `m`, `d`, `w`, `y` type suffixes)?

File: shared/api/container_snapshot.go (hunk 1 above)
Comment (monstermunchkin) [reply shared/api/container_snapshot.go]: We could just allow editing the entire snapshot config. That would make things pretty simple.

File: shared/api/container.go (hunk 2 above)
Comment (monstermunchkin) [reply shared/api/container.go]: We could just allow editing the entire snapshot config. That would make things pretty simple.

Comment (stgraber) [reply shared/api/container.go]: Yeah, showing everything and discarding everything but expiry would be fine, I think that's what we do in a few other places.

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/container.go [hunk 4]
```diff
@@ -1657,6 +1662,75 @@ func autoCreateContainerSnapshots(ctx context.Context, d *Daemon, containers []c
 	return nil
 }
 
+func pruneExpiredContainerSnapshotsTask(d *Daemon) (task.Func, task.Schedule) {
+	f := func(ctx context.Context) {
+		opRun := func(op *operation) error {
+			return pruneExpiredContainerSnapshots(ctx, d)
+		}
+
+		op, err := operationCreate(d.cluster, "", operationClassTask, db.OperationSnapshotsExpire, nil, nil, opRun, nil, nil)
```
Comment (stgraber) [inline lxd/container.go]: We should do the same thing here as we do for backup expiry, so avoid running the whole thing if there is nothing to expire, saves us noisy logging. I can send a fix for that later.

File: doc/containers.md [hunk 5]
```diff
@@ -81,6 +81,7 @@ security.syscalls.whitelist             | string    | -                 | no
 snapshots.schedule                      | string    | -                 | no            | snapshot\_scheduling                 | Cron expression (`<minute> <hour> <dom> <month> <dow>`)
 snapshots.schedule.stopped              | bool      | false             | no            | snapshot\_scheduling                 | Controls whether or not stopped containers are to be snapshoted automatically
 snapshots.pattern                       | string    | snap%d            | no            | snapshot\_scheduling                 | Pongo2 template string which represents the snapshot name (used for scheduled snapshots and unnamed snapshots)
+snapshots.expiry                        | string    | -                 | no            | snapshot\_expiry                     | Controls when snapshots are to be deleted
```
[the remark this answers is not in the document]
Comment (monstermunchkin) [reply doc/containers.md]: Done

**Pull request merged**